### PR TITLE
:seedling: Add unit tests for getDiskType, getStorageDetails, getCPUDetails, getSystemVendorDetails

### PR DESCRIPTION
**Pull request by archy-rock3t-cloud**
**What this PR does / why we need it**:

Add unit tests for getDiskType, getStorageDetails, getCPUDetails, getSystemVendorDetails

**Checklist:**

- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] E2E tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: pkg/provisioner/ironic/hardwaredetails/hardwaredetails_test.go [hunk 1]
```diff
@@ -296,3 +296,108 @@ func TestGetFirmwareDetails(t *testing.T) {
 
 	assert.Equal(t, "foobar", firmware.BIOS.Vendor)
 }
+
+func TestGetDiskType(t *testing.T) {
+	assert.Equal(t, metal3api.HDD, getDiskType(inventory.RootDiskType{Rotational: true}))
+	assert.Equal(t, metal3api.NVME, getDiskType(inventory.RootDiskType{Name: "/dev/nvme0n1"}))
+	assert.Equal(t, metal3api.NVME, getDiskType(inventory.RootDiskType{Name: "/dev/nvme0n1p1"}))
+	assert.Equal(t, metal3api.SSD, getDiskType(inventory.RootDiskType{Name: "/dev/sda"}))
```
Comment (tuminoid) [inline pkg/provisioner/ironic/hardwaredetails/hardwaredetails_test.go:304]: Convert to table-driven test.

**Review by tuminoid — COMMENTED** [review]
/ok-to-test

Thanks! Could we address some edge cases more?

1. getcpudetails invalid/empty frequency string
2. nil input to getcpudetails 
3. nil input for getstoragedetails
4. getstoragedetails with empty name but non-empty path

**Review by Rozzii — COMMENTED** [review]
This PR is failing on the linter and looks to be abandoned.

**Comment by Rozzii** [comment]
@archy-rock3t-cloud are you planning to work on this ?

File: pkg/provisioner/ironic/hardwaredetails/hardwaredetails_test.go (hunk 1 above)
Comment (archy-rock3t-cloud) [reply pkg/provisioner/ironic/hardwaredetails/hardwaredetails_test.go]: updated

**Comment by archy-rock3t-cloud** [comment]
@Rozzii yes, updated. Sorry for the late reply, the notification was lost

File: pkg/provisioner/ironic/hardwaredetails/hardwaredetails_test.go [hunk 2]
```diff
@@ -296,3 +296,171 @@ func TestGetFirmwareDetails(t *testing.T) {
...
+		{
+			Name:   "",
+			ByPath: "/dev/disk/by-path/pci-0000:00:1f.0-nvme-1",
+			Size:   256000000000,
+		},
+	})
+	assert.Len(t, storage, 1)
+	assert.Equal(t, "/dev/disk/by-path/pci-0000:00:1f.0-nvme-1", storage[0].Name)
+	assert.Equal(t, []string{"", "/dev/disk/by-path/pci-0000:00:1f.0-nvme-1"}, storage[0].AlternateNames)
+}
```
[the remark this answers is not in the document]
Comment (archy-rock3t-cloud) [reply pkg/provisioner/ironic/hardwaredetails/hardwaredetails_test.go:412]: updated

File: pkg/provisioner/ironic/hardwaredetails/hardwaredetails_test.go [hunk 3]
```diff
@@ -296,3 +296,171 @@ func TestGetFirmwareDetails(t *testing.T) {
...
+		{
+			name: "nvme partition is NVME",
+			disk: inventory.RootDiskType{Name: "/dev/nvme0n1p1"},
+			want: metal3api.NVME,
+		},
+		{
+			name: "sata device is SSD",
+			disk: inventory.RootDiskType{Name: "/dev/sda"},
+			want: metal3api.SSD,
+		},
```
[the remark this answers is not in the document]
Comment (archy-rock3t-cloud) [reply pkg/provisioner/ironic/hardwaredetails/hardwaredetails_test.go:325]: updated

File: pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go [hunk 4]
```diff
@@ -141,10 +141,13 @@ func getStorageDetails(diskdata []inventory.RootDiskType) []metal3api.Storage {
 	storage := make([]metal3api.Storage, len(diskdata))
 	for i, disk := range diskdata {
 		device := disk.Name
```
Comment (Rozzii) [inline pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go]:
```suggestion
		deviceName := disk.Name
```

File: pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go [hunk 5]
```diff
@@ -141,10 +141,13 @@ func getStorageDetails(diskdata []inventory.RootDiskType) []metal3api.Storage {
...
-		allNames := []string{device}
+		var allNames []string
+		if disk.Name != "" {
```
Comment (Rozzii) [inline pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go]:
```suggestion
		if deviceName != "" {
```

File: pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go [hunk 6]
```diff
@@ -141,10 +141,13 @@ func getStorageDetails(diskdata []inventory.RootDiskType) []metal3api.Storage {
...
+			allNames = append(allNames, disk.Name)
```
Comment (Rozzii) [inline pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go]:
```suggestion
			allNames = append(allNames, deviceName)
```

File: pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go [hunk 7]
```diff
@@ -141,10 +141,13 @@ func getStorageDetails(diskdata []inventory.RootDiskType) []metal3api.Storage {
...
+		}
 		if disk.ByPath != "" {
 			device = disk.ByPath
```
Comment (Rozzii) [inline pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go]:
```suggestion
```

**Review by Rozzii — COMMENTED** [review]
Looks good functionally, but let's rework the data access a bit. the old "device" variables has become a bit redundant after your changes.

**Review by Rozzii — COMMENTED** [review]
/approve

**Review by smoshiur1237 — COMMENTED** [review]
/lgtm
